facilitator/src/logging.rs

```diff
@@ -1,67 +1,169 @@
 /// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#LogSeverity
 #[derive(Debug, Serialize)]
 #[serde(rename_all = "UPPERCASE")]
 enum Severity {
+    Critical,
     Debug,
     Info,
     Warning,
     Error,
 }
 
-impl Severity {
-    fn from_log_level(level: Level) -> Severity {
-        match level {
+impl Display for Severity {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let string = match self {
+            Severity::Critical => "CRITICAL",
+            Severity::Debug => "DEBUG",
+            Severity::Info => "INFO",
+            Severity::Warning => "WARNING",
+            Severity::Error => "ERROR",
+        };
+        write!(f, "{}", string)
+    }
+}
+
+impl From<Level> for Severity {
+    fn from(slog_level: Level) -> Self {
+        match slog_level {
+            Level::Critical => Severity::Critical,
             Level::Error => Severity::Error,
-            Level::Warn => Severity::Warning,
+            Level::Warning => Severity::Warning,
             Level::Info => Severity::Info,
             Level::Debug => Severity::Debug,
             Level::Trace => Severity::Debug,
         }
     }
 }
 
-#[derive(Debug, Serialize)]
-#[serde(rename_all = "camelCase")]
-struct LogEntry<'a> {
-    message: String,
-    severity: Severity,
-    time: String,
-    module_path: Option<&'a str>,
-    file: Option<&'a str>,
-    line: Option<u32>,
-    thread_id: String,
+/// Options for configuring logging in this application
+pub struct LoggingConfiguration {
+    /// If true, logging output will be forced to JSON format using
+    /// [slog-json][1]. If false, logging format will be determined by detecting
+    /// whether `stderr` is a `tty`. If it is, output is formatted using
+    /// [slog-term][2]. Otherwise, `slog-json` is used.
+    ///
+    /// [1]: https://docs.rs/slog-json
+    /// [2]: https://docs.rs/slog-term
+    pub force_json_output: bool,
+    /// A version string which shall be attached to all log messages
+    pub version_string: &'static str,
+    /// Messages above this log level will be discarded
+    pub log_level: &'static str,
 }
 
-impl<'a> LogEntry<'a> {
-    fn from_record(fmt: &Formatter, record: &'a Record) -> LogEntry<'a> {
-        LogEntry {
-            severity: Severity::from_log_level(record.level()),
-            time: fmt.timestamp().to_string(),
-            message: record.args().to_string(),
-            module_path: record.module_path(),
-            file: record.file(),
-            line: record.line(),
-            thread_id: format!("{:?}", thread::current().id()),
-        }
-    }
+/// Initialize logging resources. On success, returns a root [`slog::Logger`][1]
+/// from which child loggers may be built and a GlobalLoggerGuard which must be
+/// kept live by the caller to enable the `slog_scope` global logger to function
+/// in modules that haven't yet opted into managing their own `slog::Logger`.
```

Comment:
This is a bear of a sentence tree! Can you break it up a little?

Reply:
I rewrote the doccomment. Possibly this would be even better if `setup_logging` returned a struct wrapping a `slog::Logger` and a `GlobalLoggerGuard`, and then you could have nice doccomments on that, but I will be getting rid of the `slog_scope` logger in favor of opting everything into child `slog::Loggers` very soon.